update SkiaSharp to 2.80.3

janusw opened the pull request.

janusw commented:
Unfortunately I have no idea why the builds fails here, or how to fix it (or at least work around it). None of the things I tried had any effect :(

The error that occurs is a runtime problem in the Rendering.Xaml tests:
```
1) Error : Mapsui.Rendering.Xaml.Tests.MapRendererTests.RenderBitmapAtlas
System.IO.FileLoadException : Could not load file or assembly 'System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at SkiaSharp.Utils.RentedArray`1..ctor(Int32 length)
   at SkiaSharp.SKManagedStream.OnReadManagedStream(IntPtr buffer, IntPtr size)
   at SkiaSharp.SKAbstractManagedStream.ReadInternal(IntPtr s, Void* context, Void* buffer, IntPtr size)
   at SkiaSharp.SKData.Create(SKStream stream, Int64 length)
   at SkiaSharp.SKData.Create(Stream stream, Int64 length)
   at Mapsui.Rendering.Xaml.Tests.MapRendererTests.CompareBitmaps(Stream bitmapStream1, Stream bitmapStream2, Int32 allowedColorDistance, Double proportionCorrect) in C:\projects\mapsui\Tests\Mapsui.Rendering.Xaml.Tests\MapRendererTests.cs:line 293
   at Mapsui.Rendering.Xaml.Tests.MapRendererTests.RenderBitmapAtlas() in C:\projects\mapsui\Tests\Mapsui.Rendering.Xaml.Tests\MapRendererTests.cs:line 106
```

janusw commented:
On MacOS I see warnings like this in Mapsui.Samples.Common.Desktop:

```
/Applications/Visual Studio.app/Contents/Resources/lib/monodevelop/bin/MSBuild/Current/bin/Microsoft.Common.CurrentVersion.targets(5,5): Warning MSB3277: Found conflicts between different versions of "System.Buffers" that could not be resolved.
There was a conflict between "System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" and "System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51".
    "System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" was chosen because it was primary and "System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" was not.
    References which depend on "System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" [/Users/janus/.nuget/packages/system.buffers/4.4.0/ref/netstandard2.0/System.Buffers.dll].
        /Users/janus/.nuget/packages/system.buffers/4.4.0/ref/netstandard2.0/System.Buffers.dll
          Project file item includes which caused reference "/Users/janus/.nuget/packages/system.buffers/4.4.0/ref/netstandard2.0/System.Buffers.dll".
            /Users/janus/.nuget/packages/system.buffers/4.4.0/ref/netstandard2.0/System.Buffers.dll
    References which depend on "System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" [].
        /Users/janus/.nuget/packages/harfbuzzsharp/2.6.1.7/lib/net462/HarfBuzzSharp.dll
          Project file item includes which caused reference "/Users/janus/.nuget/packages/harfbuzzsharp/2.6.1.7/lib/net462/HarfBuzzSharp.dll".
            /Users/janus/.nuget/packages/harfbuzzsharp/2.6.1.7/lib/net462/HarfBuzzSharp.dll
            /Users/janus/.nuget/packages/skiasharp.harfbuzz/2.80.2/lib/net462/SkiaSharp.HarfBuzz.dll
            /Users/janus/.nuget/packages/topten.richtextkit/0.4.139/lib/net462/Topten.RichTextKit.dll
            /Users/janus/github/Mapsui/Mapsui.Rendering.Skia/bin/Debug/netstandard2.0/Mapsui.Rendering.Skia.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.Common/bin/Debug/netstandard2.0/Mapsui.Samples.Common.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.CustomWidget/bin/Debug/netstandard2.0/Mapsui.Samples.CustomWidget.dll
        /Users/janus/.nuget/packages/skiasharp/2.80.3/lib/net462/SkiaSharp.dll
          Project file item includes which caused reference "/Users/janus/.nuget/packages/skiasharp/2.80.3/lib/net462/SkiaSharp.dll".
            /Users/janus/.nuget/packages/skiasharp/2.80.3/lib/net462/SkiaSharp.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.Common/bin/Debug/netstandard2.0/Mapsui.Samples.Common.dll
            /Users/janus/github/Mapsui/Mapsui.Rendering.Skia/bin/Debug/netstandard2.0/Mapsui.Rendering.Skia.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.CustomWidget/bin/Debug/netstandard2.0/Mapsui.Samples.CustomWidget.dll
            /Users/janus/.nuget/packages/skiasharp.harfbuzz/2.80.2/lib/net462/SkiaSharp.HarfBuzz.dll
            /Users/janus/.nuget/packages/svg.skia/0.5.0/lib/net461/Svg.Skia.dll
            /Users/janus/.nuget/packages/topten.richtextkit/0.4.139/lib/net462/Topten.RichTextKit.dll
        /Users/janus/.nuget/packages/system.memory/4.5.3/lib/netstandard2.0/System.Memory.dll
          Project file item includes which caused reference "/Users/janus/.nuget/packages/system.memory/4.5.3/lib/netstandard2.0/System.Memory.dll".
            /Users/janus/.nuget/packages/system.memory/4.5.3/lib/netstandard2.0/System.Memory.dll
            /Users/janus/.nuget/packages/harfbuzzsharp/2.6.1.7/lib/net462/HarfBuzzSharp.dll
            /Users/janus/.nuget/packages/skiasharp/2.80.3/lib/net462/SkiaSharp.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.Common/bin/Debug/netstandard2.0/Mapsui.Samples.Common.dll
            /Users/janus/github/Mapsui/Mapsui.Rendering.Skia/bin/Debug/netstandard2.0/Mapsui.Rendering.Skia.dll
            /Users/janus/github/Mapsui/Samples/Mapsui.Samples.CustomWidget/bin/Debug/netstandard2.0/Mapsui.Samples.CustomWidget.dll
            /Users/janus/.nuget/packages/skiasharp.harfbuzz/2.80.2/lib/net462/SkiaSharp.HarfBuzz.dll
            /Users/janus/.nuget/packages/svg.custom/0.5.0/lib/net461/Svg.Custom.dll
            /Users/janus/.nuget/packages/svg.model/0.5.0/lib/net461/Svg.Model.dll
            /Users/janus/.nuget/packages/svg.skia/0.5.0/lib/net461/Svg.Skia.dll
            /Users/janus/.nuget/packages/topten.richtextkit/0.4.139/lib/net462/Topten.RichTextKit.dll (MSB3277) (Mapsui.Samples.Common.Desktop)
```

pauldendulk commented:
> I don't think this will help. The failures appeared directly after the first commit on this branch, which only updates SkiaSharp, without removing any nugets (the latter was only done later in an attempt to get rid of the failures).

ah right, so apparently some dependencies where changed in SkiaSharp.

> Instead, maybe it helps to also add the System.Buffers package explicitly? Will try that soon.

Yes, we should try that. Also you have to get the version right. In my experience usually other conflicts pop up after that and you need to continue this process for a while.

pauldendulk commented:
If you go this route the first thing to do is to remove all unnecessary nuget references in the project and all its dependencies.

pauldendulk commented:
It may even be worth it to start with an new empty .csproj and add everything that is needed step by step.

janusw commented:
> > Instead, maybe it helps to also add the System.Buffers package explicitly? Will try that soon.
> 
> Yes, we should try that. Also you have to get the version right.

I tried in 6fdcc79 to add System.Buffers explicitly in version 4.5.0, which should correspond to assembly version 4.0.2.0 (as mentioned in the error message). I also had to remove System.Memory in a couple of places to avoid conflicts.

Still, the errors persist :/

So maybe I should try to do some dependency cleanup first, as you suggest. Will try that soon.

inforithmics commented on Mapsui.UI.Wpf/Mapsui.UI.Wpf.csproj:
```diff
@@ -11,10 +11,11 @@
     <PackageReference Include="BruTile" Version="4.0.0" />
-    <PackageReference Include="SkiaSharp" Version="2.80.2" />
+    <PackageReference Include="SkiaSharp" Version="2.80.3" />
+    <PackageReference Include="SkiaSharp.Views.WPF" Version="2.80.3" />
```
I think this was not intended to add a reference to SkiaSharp.Views.WPF here

janusw replied:
Ah, very good point indeed! I did not actually notice that before, and I'm not sure how it happened.

IIRC I did the package update via the VS UI, so I assume the fact that the target-framework condition was not treated properly is some kind of VS bug.

Thanks for catching this! I will fix it and check if that cures it (I assume it will).

inforithmics commented on Mapsui.UI.Wpf/Mapsui.UI.Wpf.csproj:
```diff
@@ -11,10 +11,11 @@
+    <PackageReference Include="SkiaSharp.Views" Version="2.80.3" />
   </ItemGroup>
   <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp3.1' ">
     <PackageReference Include="SkiaSharp.Views.WPF" Version="2.80.2" />
```
Here should teh SkiaSharp.Views.Wpf Component be upgraded to 2.80.3

inforithmics commented:
I run the Unit Tests locally. After i fixed the Problem I mentioned in the comment. And removed 
System.Buffers and System.Runtime.CompilerServices.Unsafe from all mapsui projects the Unit tests run through and it compiled.
Because in my opinion they are not needed, And when the dependencies bring in the right Version.

inforithmics commented on Mapsui.Rendering.Xaml/Mapsui.Rendering.Xaml.csproj:
```diff
@@ -29,8 +29,6 @@
-    </PackageReference>
     <PackageReference Include="System.Net.Http" Version="4.3.4">
     </PackageReference>
     <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.7.1">
```
All occurences should be removed. (I don't comment on the other occurences)

inforithmics commented on Tests/Mapsui.Rendering.Xaml.Tests/Mapsui.Rendering.Xaml.Tests.csproj:
```diff
@@ -73,11 +73,10 @@
-    <PackageReference Include="SkiaSharp" Version="2.80.2">
+    <PackageReference Include="SkiaSharp" Version="2.80.3">
     </PackageReference>
+    <PackageReference Include="System.Buffers" Version="4.5.0" />
```
System Buffers should be removed. (I don't comment on the other occurences)

janusw commented:
> I run the Unit Tests locally. After i fixed the Problem I mentioned in the comment. And removed
> System.Buffers and System.Runtime.CompilerServices.Unsafe from all mapsui projects the Unit tests run through and it compiled.

Yes, I can confirm this (both locally and via AppVeyor). Thanks a lot for your help!

inforithmics commented on Tests/Mapsui.Rendering.Skia.Tests/Mapsui.Rendering.Skia.Tests.csproj:
```diff
@@ -12,22 +12,22 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
-    <OutputPath>bin\x64\Debug\</OutputPath>
+    <OutputPath>bin\x64\Debug\net48\net48</OutputPath>
```
I think one net48 is sufficent it doesn't need to be net48\net48

janusw replied:
Yeah, this is weird. Apparently this was changed automatically when updating packages via VS. The first `net48` was added in the first commit (when updating SkiaSharp), the second in the second commit (when updating Svg.Skia). I will clean this up.

janusw replied:
Fixed. I simply removed all of these OutputPath changes, which are not related to the package updates.

inforithmics commented on Tests/Mapsui.Rendering.Skia.Tests/Mapsui.Rendering.Skia.Tests.csproj:
```diff
@@ -12,22 +12,22 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
-    <OutputPath>bin\x86\Debug\</OutputPath>
+    <OutputPath>bin\x86\Debug\net48\net48</OutputPath>
```
I think one net48 is sufficent it doesn't need to be net48\net48

inforithmics commented on Tests/Mapsui.Rendering.Skia.Tests/Mapsui.Rendering.Skia.Tests.csproj:
```diff
@@ -12,22 +12,22 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
-    <OutputPath>bin\x86\Release\</OutputPath>
+    <OutputPath>bin\x86\Release\net48\net48</OutputPath>
```
I think one net48 is sufficent it doesn't need to be net48\net48

janusw commented:
Merged. Thanks again for the support and the review, @inforithmics!
